File: 프로그래머스/3/43162. 네트워크/네트워크.java
import java.util.*;
class Solution {
    ArrayList<Integer>[] graph;
    boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;
        graph = new ArrayList[n+1];
        visited = new boolean[n+1];
        for(int i=0;i<n;i++){
            graph[i] = new ArrayList<>();
        }
        
        for(int i=0;i<computers.length;i++){
            for(int j=0;j<computers.length;j++){
                if(computers[i][j]==1){
                    graph[i].add(j);
                    graph[j].add(i);
                }
            }
        }
        
        for(int i=0;i<n;i++){
            if(!visited[i]){
                dfs(i);
                answer++;
            }
        }
        
        return answer;
    }
    
    public void dfs(int v){
        visited[v]=true;
        for(int next : graph[v]){
            if(!visited[next])dfs(next);
        }
    }
    
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;
bool visited[201];

void dfs(vector<vector<int>>computers, int num){
    for(int i=0;i<computers[num].size();i++){
        if(!visited[i]&&computers[num][i]==1){
            visited[i]=true;
            dfs(computers,i);
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    for(int i=0;i<computers.size();i++){
        for(int j=0;j<computers[i].size();j++){
            if(i==j){
                continue;
            }
            if(!visited[j]&&computers[i][j]==1){
                visited[j]=true;
                visited[i]=true;
                dfs(computers,j);
                answer++;
            }
        }
    }
    for(int i=0;i<computers.size();i++){
        if(!visited[i])answer++;
    }
    return answer;
}